### PR TITLE
BFD-2712: Simplify pipeline app

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[BFD-2712](https://jira.cms.gov/browse/BFD-2712)

**User Story or Bug Summary:**

As a BFD contributor I want the pipeline to be less complex so that I can better maintain it.

---

### What Does This PR Do?

Most of the file changes are trivial.  The biggest changes are in `PipelineJobRunner` and `PipelineManager`.

* Adds `PipelineJobRunner` to serve as a wrapper around jobs which takes care of lifecycle events like scheduling, starts, stops, and error handling.
* Modifies `PipelineManager` to manage a simple pool of `PipelineJobRunner`s.  One per job.  Tracking lifecycle events received from them and gracefully managing application shutdown when all jobs are complete or any throw an exception.  Triggers graceful shutdown by all jobs if any of them fail with an exception.
* Removes unused generic `PipelineJobArguments` on `PipelineJob` since it introduced complexity for no gain.
* Eliminates many classes made obsolete by the simpler design.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/load/RifLoader.java [hunk 1]
```diff
@@ -294,23 +311,18 @@ record -> {
...
-     * on this thread to kill the pipeline. Exceptions thrown from the batch thread with the
-     * error are ignored/lost, so have this synchronized boolean signal to do this specifically on the main thread.
-     * We also can't use a passed-in error handler from up the execution chain or else the pipeline manager shutdown procedure
-     * waits for this thread to shut down, which is waiting on the above 72-hour timeout before it unblocks.
-     *
-     * FUTURE: If we wish to be more discerning about recoverable errors, we could return false here which will
-     * allow the pipeline to continue operating if an error is considered non-fatal (it also currently moves the failed files
-     * to "Done" instead of "Failed", which would need to be fixed.)
+     * If any batch load tripped a fatal error, we call the error handler.
+     * The stream will have already cleanly halted processing.
```
Comment (lsmitchell) [inline apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/load/RifLoader.java:320]: If part of the error handler process is waiting for all processing threads to terminate, as it did in the past, it'll sit there waiting for the full timeout above on awaitTermination (which is set to multiple days to allow the loads to run in a long-long scenario). Can't discern by looking at the new flow if that will be a problem. When I was testing this scenario to do the bugfix that added fatalFailure, I had two test cases I (manually) ran to test the "hung pipeline" scenario; one where a rif file had an invalid data item in a row (say, a number that was too big to fit in int) and one where the header column name was bad (messing up the name of a header.) 

In retrospect I probably should have written an actual test to cover that case, but that'd be something it'd be good to make sure can't happen so we don't revive https://jira.cms.gov/browse/BFD-2329

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineJobRunner.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    }
+  }
+
+  /**
+   * Runs the job once and reports its outcome to the {@link Tracker}.
+   *
+   * @throws Exception passed through if the job terminates with an exception
+   */
+  private void runJob() throws Exception {
+    final var id = tracker.beginningRun(job);
```
Comment (lsmitchell) [inline apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineJobRunner.java:69]: Here's a better example of where var adds ambiguity. I'm reading through the review, and now I have to ask myself what object type is this? I can go through the effort to go find what the method returns, but it seems like needless obfuscation that puts the onus on the reader to go sleuth out. (Also something easier to figure out in IDE, but hard to read in a review)

**Review by lsmitchell — COMMENTED** [review]
Looks like a solid improvement by removing a lot of needless complication with the many many classes pipeline used to parcel out its work.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/DefaultDataSetMonitorListener.java [hunk 3]
```diff
@@ -62,6 +63,7 @@ public void dataAvailable(RifFilesEvent rifFilesEvent) {
                     PipelineApplication.class.getSimpleName(), "dataSet", "processed"))
             .time();
 
+    final var failed = new AtomicBoolean(false);
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/DefaultDataSetMonitorListener.java:66]: Implicit typing when the type is obvious does not harm readability.  For example, how is this any clearer?

```
PGSimpleDataSource dataSource = new PGSimpleDataSource();
```

Java is often knocked for verbosity and small reductions seem like a win to me.  The introduction of `var` is just java catching up to more modern language design practice.

This particular case is perfectly obvious what the type is.  A stronger case could be made when the initialization is via a method call other than a constructor that does not convey the type information.

File: apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/PipelineJobRunnerTest.java [hunk 4]
```diff
@@ -0,0 +1,380 @@
+package gov.cms.bfd.pipeline.app;
+
+import static org.junit.jupiter.api.Assertions.*;
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/PipelineJobRunnerTest.java:3]: Hmm, my IDEA settings seem to have regressed.  I'll fix this.

File: apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/PipelineJobRunnerTest.java [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+   */
+  @Test
+  void runsUntilStopped() throws Exception {
+    final var repeatMills = 5_000L;
+    doReturn(Optional.of(new PipelineJobSchedule(repeatMills, ChronoUnit.MILLIS)))
+        .when(job)
+        .getSchedule();
+
+    // job can run twice before being told to stop
+    doReturn(true, true, true, true, false).when(tracker).jobsCanRun();
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/PipelineJobRunnerTest.java:204]: I'll improve the comment.

File: apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/PipelineJobRunnerTest.java [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+   */
+  @Test
+  void runsUntilInterruptedBetweenRuns() throws Exception {
+    final var repeatMills = 5_000L;
+    doReturn(Optional.of(new PipelineJobSchedule(repeatMills, ChronoUnit.MILLIS)))
+        .when(job)
+        .getSchedule();
+
+    // job can run twice before being told to stop
+    doReturn(true).when(tracker).jobsCanRun();
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/PipelineJobRunnerTest.java:254]: Good catch.  That was a copy-paste holdover.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineJobRunner.java (hunk 2 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineJobRunner.java:69]: I don't see much sleuthing required here but I'll make the change.

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/load/RifLoader.java (hunk 1 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/load/RifLoader.java:320]: Good question.  Previously the batch processing was stream was not tied into the error handling logic so it only terminated when it hit end of file.

This change in `RifLoader` (and same right below it for alternative stream) stops parsing/queuing batches once an exception has been reported:

```
            .takeWhile(record -> error.get() == null) // stop if an exception is thrown
```

Likewise this change in `DefaultDataSetMonitorListener` does the same at the manifest level.

File: apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/PipelineJobRunnerTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/PipelineJobRunnerTest.java]: Fixed.

File: apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/PipelineJobRunnerTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/PipelineJobRunnerTest.java]: Expanded/fixed all of the comments in test methods.

File: apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/PipelineJobRunnerTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/PipelineJobRunnerTest.java]: Expanded/fixed all of the comments in test methods.

**Review by mackec1 — APPROVED** [review]

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/load/RifLoader.java (hunk 1 above)
Comment (lsmitchell) [reply apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/load/RifLoader.java:320]: Thanks! Seems the new flow will take care of this. I meant to mention above, but we can probably remove fatalFailure boolean then, I think it's been obsoleted.

**Review by lsmitchell — APPROVED** [review]
Thanks for the adjustments. Noted one thing that can be optionally removed (fatalFailure) since I think it's useless now.

**Comment by brianburton** [comment]
> Thanks for the adjustments. Noted one thing that can be optionally removed (fatalFailure) since I think it's useless with the new flow.

Great catch.  I have removed the unused field.

**Review by mjburling — APPROVED** [review]

**Review by mackec1 — APPROVED** [review]

**Pull request merged**